lib/jetstream/pull_consumer.ex

```diff
@@ -1,42 +1,176 @@
 defmodule Jetstream.PullConsumer do
+  @moduledoc """
+  A behaviour which provides the NATS JetStream Pull Consumer functionalities.
+
+  When a Consumer is pull-based, it means that the messages will be delivered when the server is asked
+  for them.
+
+  ## Usage
+
+  Declare a module which uses `Jetstream.PullConsumer` and implements the `handle_message/1` function.
+
+  ```
+  defmodule MyApp.PullConsumer do
+    use Jetstream.PullConsumer
+
+    def handle_message(message) do
+      # Do some processing with the message.
+      :ack
+    end
+  end
+  ```
+
+  You can then place your Pull Consumer in a supervision tree. Remember that you need to have the
+  `Gnat.ConnectionSupervisor` set up.
+
+  ```
+  defmodule MyApp.Application do
+    use Application
+
+    @impl true
+    def start(_type, _args) do
+      children = [
+        # Create NATS connection
+        {Gnat.ConnectionSupervisor, ...},
+        # Start NATS Jetstream Pull Consumer
+        {MyApp.PullConsumer,
+          %{
+            connection_name: :gnat,
+            stream: %Jetstream.API.Stream{name: "TEST_STREAM", subjects: "test"},
+            consumer: %Jetstream.API.Consumer{stream_name: "TEST_STREAM", name: "TEST_CONSUMER"}
+          }}
+      ]
+      opts = [strategy: :one_for_one]
+      Supervisor.start_link(children, opts)
+    end
+  end
+  ```
+
+  The above will create a new stream and consumer. You can use existing stream/consumer by providing its
+  name:
+
+  ```
+  %{
+    connection_name: :gnat,
+    stream: "TEST_STREAM",
+    consumer: "TEST_CONSUMER"
+  }
+  ```
+  """
+
   use GenServer
 
+  @doc """
+  Called when the Pull Consumer recives a message. Depending on the value it returns, the acknowledgement
+  is or is not sent.
+
+  Possible return values explained:
+
+  * `:ack` - acknowledges the message was handled and requests delivery of the next message to the reply
+    subject.
+
+  * `:nack` - signals that the message will not be processed now and processing can move onto the next
+    message, NAK'd message will be retried.
+
+  * `:noreply` - nothing is sent.
+
+  The client should be listening for acknowledgements on the `$JS.ACK.<stream>.<consumer>.>` subject.
+  """
+  @callback handle_message(message :: Jetstream.message()) ::
+              :ack | :nack | :noreply
+
+  @type settings :: %{
+          connection_name: pid() | atom(),
+          stream_name: binary() | Jetstream.API.Stream.t(),
+          consumer_name: binary() | Jetstream.API.Consumer.t(),
+          module: atom()
+        }
+
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour Jetstream.PullConsumer
+
+      @type settings :: %{
+              connection_name: pid() | atom(),
+              stream: binary() | Jetstream.API.Stream.t(),
+              consumer: binary() | Jetstream.API.Consumer.t()
+            }
+
```

Comment:
```suggestion
      @doc """
      Returns a specification to start this module under a supervisor.
      See `Supervisor`.
      """
```

https://github.com/elixir-lang/elixir/blob/v1.13.3/lib/elixir/lib/gen_server.ex#L746-L752